### PR TITLE
Feature: vHost templates (version 2)

### PR DESCRIPTION
**Pull request by Churro**
@hpmewes presented a nice implementation of pre-definable vHost templates in PR #233 .
Presumably due to some unresolved issues, his implementation has not been merged finally and remained as a pending pull request.

In this PR, a reworked implementation is provided with:
- bugfixes all over
- added support for apache and lighttpd also
- added an update sequence (instead of only modifying froxlor.sql)
- added english language variables
- added missing parts in admin_vhostsettings.php
- added parameter replacements as available since PR #244 

The solution has been tested working on different setups and should integrate / upgrade flawlessly.

### CODE REVIEW

**Comment by d00p** [comment]
Instead of posting a PR every 2 hours, why don't you just join us on IRC and we talk about all your ideas and stuff...it's a bit stressful :)

File: admin_vhostsettings.php [hunk 1]
```diff
@@ -0,0 +1,233 @@
+<?php
+
+/**
+ * This file is part of the Froxlor project.
+ * Copyright (c) 2003-2009 the SysCP Team (see authors).
```
Comment (d00p) [inline admin_vhostsettings.php]: new files do not have anything to do with SysCP

File: admin_vhostsettings.php [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+ * Copyright (c) 2010 the Froxlor Team (see authors).
+ *
+ * For the full copyright and license information, please view the COPYING
+ * file that was distributed with this source code. You can also view the
+ * COPYING file online at http://files.froxlor.org/misc/COPYING.txt
+ *
+ * @copyright  (c) the authors
+ * @author     Florian Lippert <flo@syscp.org> (2003-2009)
```
Comment (d00p) [inline admin_vhostsettings.php]: new files do not have anything to do with SysCP

File: lng/english.lng.php [hunk 3]
```diff
@@ -1969,4 +1969,21 @@
...
+$lng['admin']['vhostsettings']['vhostsettings'] = 'Predefined VHost configuration';
+$lng['admin']['vhostsettings']['novhostsettings'] = 'No predefined VHost configuration';
+$lng['admin']['vhostconfig']['socket_dir'] = 'Will be replaced with the corresponding Unix socket path, e.g. /var/run/domain-php-fpm.socket.';
+$lng['admin']['vhostconfig']['docroot'] = 'Will be replaced with the DocumentRoot.';
+$lng['admin']['vhostconfig']['ip'] = 'Will be replaced with the IP address of the Vhost.';
+$lng['admin']['vhostconfig']['port'] = 'Will be replaced with the port number of the Vhost.';
+$lng['admin']['vhostconfig']['scheme'] = 'Will be set to <strong>http</strong> or <strong>https</strong> in case SSL / TLS is activated for the Vhost.';
+$lng['admin']['vhostconfig']['vhost_usedefaultlocation']['title'] = 'Use the standard Vhost location directives';
+$lng['admin']['vhostconfig']['vhost_usedefaultlocation']['desc'] = 'If activated, the standard Vhost location directives, such as <br>location / { ... }<br> location @rewrites { ... }<br> location ~ ^(.+?\.php)(/.*)?$ { ... }<br> are excluded from the Vhost configuration file.';
```
Comment (d00p) [inline lng/english.lng.php]: nginx specific...

Comment (Churro) [inline lng/english.lng.php]: In this case, "such as" is used as a synonym for "for example". Of course, you could mention the default directives of all three webservers here but that would probably miss the point.

**Comment by Churro** [comment]
Alright, Copyright fixed. Any other issues?

File: lng/english.lng.php (hunk 3 above)
Comment (d00p) [inline lng/english.lng.php]: Well even "vhost location directives" is nginx specific....no apache-user knowns what a location-directive is (generally speaking). Something more abstract and "valid" for all webservers would be nice (imho)

Comment (Churro) [inline lng/english.lng.php]: While looking for a more generalized description, I noticed that it actually makes no sense to show this option to Apache / lighttpd users at all. Aside from nginx, no default diroptions are included / required anyway so it is somehow redundant to let the user disable it.

Summarizing, commit 373cb91 now simply hides this option for webservers other than nginx and sets it to the default value in the db (--> thereby, retaining the option that the admin may change the webserver occasionally). The commit also includes a typo fix and fixes the indentation of the SQL statement in the db upgrade step.

**Comment by nachtgeist** [comment]
This will lead to serious confusion when the webserver is switched as this might lead to e.g. apache config directives ending up in ngingx server-configs.

The easiest way to fix this would be to provide separate textareas for each supported webserver and in the cron scripts only to incoroporate the contents from the appropriate column.

**Comment by EleRas** [comment]
But this is already the case with the "default vhostsettings", they break as soon as you switch webservers

**Comment by nachtgeist** [comment]
Good point, didn't realize that as I never used that myself.

Then they need be split up per webserver as well.

**Comment by d00p** [comment]
I'd suggest to add a field 'Webserver' to the table which leads to a select box and then you can create templates for whatever Webserver and they are only used if the templates Webserver matches the currently used...

**Comment by Churro** [comment]
Alright, this also matches with what I just proposed in IRC. A PR is going to follow soon.

**Pull request merged**